torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3000,8 +3015,13 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     # would thus be in this reduced precision, or
                     # 2) parameters did not have precision reduced, but grads
                     # had reduced precision for communication.
+                    # If a lower precision hook is registered, gradients are casted
+                    # back by the hook.
+                    # However, if a low precision hook is attached to the model.
+                    # casting happens inside the hook.
                     if (
-                        self._mixed_precision_enabled_for_params() or self._mixed_precision_enabled_for_reduce()
+                        not self._low_precision_hook_enabled() and
```

Comment:
We should think about how to refactor this, such as, if reduce_dtype is available, then just register a reduced precision communication hook, but that requires a bit more flushing out. Might be good to file a follow up issue.

Reply:
I will create a follow-up issue after this PR is landed

Reply:
We check `not self._low_precision_hook_enabled` in L2980, so this check should always be true within right?

Reply:
@rohan-varma right, thanks for catching this

Reply:
@rohan-varma after looking into this: first check on now L3024 is for pre-casting gradients for all cases. This line is for re-casting gradients back into full-precision for `NO_SHARD`. This does not belong to the body of L3023 if, so this check is still needed. Otherwise, `comm_hook` will cast grads back and will occupy memory for `orig_param_grad_data` and after comm_hook is done, without this check here, it will attempt a second re-casting

Reply:
Sounds good